### PR TITLE
335 updates to externalPropertyFiles  [https://github.com/oasis-tcs/sarif-spec/issues/335]

### PR DESCRIPTION
**Pull request by harleenkohli**
changes already in master:
- In the externalPropertyFiles object:
    * Rename the properties property to externalizedProperties.

changes in this PR:
- In the externalPropertyFiles object:
    * Rename the properties property to externalizedProperties.
    * Restore the properties property, which now refers to the object level property bag, not to externalized content.
    * In the externalPropertyFiles object:
    * Rename the instanceGuid property to guid.
    * Rename the artifactLocationproperty to location

### CODE REVIEW

File: src/Sarif/Autogenerated/SarifRewritingVisitor.cs [hunk 1]
```diff
@@ -330,6 +330,14 @@ public virtual ExternalPropertyFiles VisitExternalPropertyFiles(ExternalProperty
...
                 }
 
+                if (node.ThreadFlowLocations != null)
+                {
+                    for (int index_0 = 0; index_0 < node.ThreadFlowLocations.Count; ++index_0)
+                    {
+                        node.ThreadFlowLocations[index_0] = VisitNullChecked(node.ThreadFlowLocations[index_0]);
+                    }
+                }
+
```
Comment (harleenkohli) [inline src/Sarif/Autogenerated/SarifRewritingVisitor.cs]: this is strictly not part of this PR, but when i regenerated this file i noticed this additional check and thought of bringing is to attention in case it was mistakenly not included earlier?

File: src/Sarif/Autogenerated/SarifRewritingVisitor.cs [hunk 2]
```diff
@@ -805,6 +813,13 @@ public virtual Run VisitRun(Run node)
                 }
 
                 node.ExternalPropertyFiles = VisitNullChecked(node.ExternalPropertyFiles);
+                if (node.ThreadFlowLocations != null)
+                {
+                    for (int index_0 = 0; index_0 < node.ThreadFlowLocations.Count; ++index_0)
+                    {
+                        node.ThreadFlowLocations[index_0] = VisitNullChecked(node.ThreadFlowLocations[index_0]);
+                    }
+                }
```
Comment (harleenkohli) [inline src/Sarif/Autogenerated/SarifRewritingVisitor.cs]: same comment as above: this is strictly not part of this PR, but when i regenerated this file i noticed this additional check and thought of bringing is to attention in case it was mistakenly not included earlier?

File: src/Sarif/Autogenerated/SarifRewritingVisitor.cs [hunk 3]
```diff
@@ -831,6 +846,14 @@ public virtual SarifLog VisitSarifLog(SarifLog node)
...
                     }
                 }
+
+                if (node.InlineExternalPropertyFiles != null)
+                {
+                    for (int index_0 = 0; index_0 < node.InlineExternalPropertyFiles.Count; ++index_0)
+                    {
+                        node.InlineExternalPropertyFiles[index_0] = VisitNullChecked(node.InlineExternalPropertyFiles[index_0]);
+                    }
+                }
```
Comment (harleenkohli) [inline src/Sarif/Autogenerated/SarifRewritingVisitor.cs]: same comment as above: this is strictly not part of this PR, but when i regenerated this file i noticed this additional check and thought of bringing is to attention in case it was mistakenly not included earlier?

**Comment by harleenkohli** [comment]
>             "$ref": "#/definitions/threadFlowLocation"

this is the only property which s not of type externalPropertyFile, felt weird given the pattern and the description. Is this valid?

---
Refers to: src/Sarif/Schemata/sarif-schema.json:539 in 217ae3e. [](commit_id = 217ae3e882d5952a46a23051f7dae4d91ec61a00, deletion_comment = False)

File: src/Sarif.UnitTests/TestData/PrereleaseCompatibilityTransformer/Inputs/WithExternalPropertyFiles.sarif [hunk 4]
```diff
@@ -0,0 +1,56 @@
+﻿{
+    "$schema": "http://json.schemastore.org/sarif-2.0.0-csd.2.beta.2018-10-10",
+    "version": "2.0.0-csd.2.beta.2018-10-10",
+    "runs": [{
+      "tool": {
+        "name": "CodeScanner",
+        "semanticVersion": "2.1.0"
+      },
+      "results": [],
+      "externalPropertyFiles" : {
```
Comment (harleenkohli) [inline src/Sarif.UnitTests/TestData/PrereleaseCompatibilityTransformer/Inputs/WithExternalPropertyFiles.sarif]: >"externalPropertyFiles" : { [](start = 6, length = 27)

there is no test case which tests this code path so i generated a net new test, i have no idea if any of these values resemble a real world so please cross validate if this can be improved.

File: src/Sarif/Autogenerated/SarifRewritingVisitor.cs [hunk 5]
```diff
@@ -330,6 +330,14 @@ public virtual ExternalPropertyFiles VisitExternalPropertyFiles(ExternalProperty
                     }
                 }
 
+                if (node.ThreadFlowLocations != null)
+                {
+                    for (int index_0 = 0; index_0 < node.ThreadFlowLocations.Count; ++index_0)
+                    {
+                        node.ThreadFlowLocations[index_0] = VisitNullChecked(node.ThreadFlowLocations[index_0]);
+                    }
+                }
```
Comment (harleenkohli) [inline src/Sarif/Autogenerated/SarifRewritingVisitor.cs]: this is strictly not part of this PR, but when i regenerated this file i noticed this additional check and thought of bringing is to attention in case it was mistakenly not included earlier?

File: src/Sarif/Autogenerated/SarifRewritingVisitor.cs (hunk 1 above)
Comment (harleenkohli) [inline src/Sarif/Autogenerated/SarifRewritingVisitor.cs]: this is strictly not part of this PR, but when i regenerated this file i noticed this additional check and thought of bringing is to attention in case it was mistakenly not included earlier?

File: src/Sarif/Autogenerated/SarifRewritingVisitor.cs [hunk 6]
```diff
@@ -912,6 +935,8 @@ public virtual ToolComponent VisitToolComponent(ToolComponent node)
         {
             if (node != null)
             {
+                node.ShortDescription = VisitNullChecked(node.ShortDescription);
+                node.FullDescription = VisitNullChecked(node.FullDescription);
```
Comment (harleenkohli) [inline src/Sarif/Autogenerated/SarifRewritingVisitor.cs]: same comment as above: this is strictly not part of this PR, but when i regenerated this file i noticed this additional check and thought of bringing is to attention in case it was mistakenly not included earlier?

**Comment by harleenkohli** [comment]
>                         "$ref": "#/definitions/threadFlowLocation"

this is the only property which s not of type externalPropertyFile, felt weird given the pattern of the rest of the parent object and the description. Is this valid?

---
Refers to: src/Sarif/Schemata/sarif-schema.json:539 in 4321eb9. [](commit_id = 4321eb95e075dcdf8212afcf16fa0a7e87e3edf1, deletion_comment = False)

File: src/Sarif.UnitTests/TestData/PrereleaseCompatibilityTransformer/Inputs/WithExternalPropertyFiles.sarif [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        "conversion": {
+            "artifactLocation": {
+                "uri": "Test/Test.txt",
```
Comment (michaelcfanning) [inline src/Sarif.UnitTests/TestData/PrereleaseCompatibilityTransformer/Inputs/WithExternalPropertyFiles.sarif]: >Test/Test.txt [](start = 24, length = 13)

name this 'externalized/conversion.sarif-external-properties'. no uriBaseId. remove itemCount for anything that's not an array

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs [hunk 8]
```diff
@@ -304,6 +306,79 @@ private static void UpdateToolObjectToolComponentProperties(JObject tool)
             }
         }
 
+        private static void UpdateAllExternalPropertyFilePropertyTypes(JObject run)
```
Comment (michaelcfanning) [inline src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: >UpdateAllExternalPropertyFilePropertyTypes [](start = 28, length = 42)

i think all this code could be replaced by retrieving the run.externalPropertyFiles, then use the helper method to globally rename things.

RecursivePropertyRename(externalPropertyFiles, universallyRenamedMembers);

File: src/Sarif/Autogenerated/SarifRewritingVisitor.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/Sarif/Autogenerated/SarifRewritingVisitor.cs]: yes! this was a bug in the SDK! the visitor wouldn't visit newly added things.

---
In reply to: [262333747](https://github.com/Microsoft/sarif-sdk/pull/1328#discussion_r262333747) [](ancestors = 262333747)
